Migrate Posts List to to swift

diegoreymendez opened the pull request:
Migrates PostListViewController to Swift.  Since porting just a part of the code is not feasible, I'm describing below how this'll be taken care of.

**Porting and reviewing guidelines:**

First off I've set some porting rules for myself, in order to make the work easier.  Please keep these in mind when reviewing the code:

- The ported code doesn't need to be better, just equivalent.  Improvements can be made after this initial porting step is complete.
- Additional comments won't be added.  Documenting methods is outside the scope of this particular PR.
- Make sure the code is 1:1 as much as possible (saving the occasional improvement).

**Code Reviews:**

I've split the review in sections (the `#pragma mark` sections from the original ObjC file), so that the reviewers can focus on smaller parts of the whole thing.

Please review @jleandroperez 
- [x] Header File
- [x] Constants & Properties
- [x] Lifecycle Methods
- [x] Configuration

Please review @SergioEstevao  
- [x] Sync Methods
- [x] Actions
- [x] TableView Handler Delegate Methods
- [x] Table View Handling

Please review @aerych 
- [ ] Post Actions
- [ ] Filter related
- [ ] Cell Delegate Methods

**Testing:**

Please make sure to run these tests both with WP.com and self-hosted.

Steps to find this VC in the app:

1. Go to My Sites (first icon with the WordPress logo in the tab bar)
2. Select a site
3. Tap "Blog Posts"... et voila!

Tests for @jleandroperez:

- [x] Test 1:
1. Make sure you open a blog without posts.
2. Make sure that even if you select any filter from the top, no posts are shown on screen.
3. Make sure that each section is showing an appropriate message at the center of the screen (you'll know when you read them!).


- [x] Test 2:
1. Make sure you open a blog without posts.
2. Tap the search icon.
3. Type in any word.
4. Make sure the text at the center of the screen shows an appropriate message for the filter you've selected.

Tests for @SergioEstevao:

- [ ] Test 3:
Same as test 2 but with results.  Make sure the results are shown properly.

- [ ] Test 4:
1. Add posts through the blog's web page.
2. Make sure at least 1 of them is published, 1 draft, 1 scheduled and 1 trashed.
3. Check in the app to make sure the posts are in the appropriate section.

Tests for @aerych:

- [ ] Test 5:
1. Same as test 4 but once you have the posts...
2. Move them around or delete them using WPiOS, and make sure the changes are properly reflected in the web interface.


- [ ] Test 6:
1. Create a new post in WPiOS for each one of these: published, draft, scheduled.
2. Make sure they show up properly and that they're uploaded to your blog.

And any other test you can think of in that screen. :)

**UPDATE:** I'm reassigning @bummytime 's part of the review to @SergioEstevao, since Matt is on vacation next week.

diegoreymendez commented:
I'll list here the dependencies that need to be merged before this PR is ready for review, as I'm breaking it up a bit:

- https://github.com/wordpress-mobile/WordPress-iOS/pull/5158

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    class func controllerWithBlog(blog: Blog) -> PostListViewController {
+        
+        let storyBoard = UIStoryboard(name: "Posts", bundle: NSBundle.mainBundle())
+        let controller = storyBoard.instantiateViewControllerWithIdentifier("PostListViewController") as! PostListViewController
```
I've got second thoughts on whether we wanna stick to a Storyboard, or not. There are just two viewControllers there, and (if we) switched to Nibs, this method could just turn into an initializer.

But that's definitely another issue! =D

diegoreymendez replied:
Even if there is space for improvements we should keep this PR focused on the migration in order to avoid breaking other stuff.

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    
+    static private let statsCacheInterval = NSTimeInterval(300) // 5 minutes
+    static private let postCardEstimatedRowHeight = Float(100.0)
+    static private let postCardRestoreCellRowHeight = Float(54.0)
```
Maybe just store these two constants as `CGFloat` instead of `Float`, since they're casted anyways whenever they're used

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    
+    private func noResultsTitlesWhenSearching() -> [PostListStatusFilter:String] {
+        
+        let draftMessage = String(format: NSLocalizedString("No drafts match your search for %@", comment: "The '%@' is a placeholder for the search term."), currentSearchTerm()!)
```
`<Annoying+Nitpticky>` some methods have a spacing between the signature and first instruction, others don't. `</Annoying+Nipticky>`

diegoreymendez replied:
I've changed it in this line not to have an extra space.

However there's a reason why the spacing is not always the same: I always try to use the spacing to increase readability - if you check the method now you'll notice that the method definition blends in with the following block of code.  The idea behind this spacing was to make the code easier to read.

jleandroperez replied:
In that case, maybe... we should move the opening brace `{` to a new line, always.

That way everything would be consistent, no matter what.

diegoreymendez replied:
All `{`s are in the same line as methods definition.  That's pretty consistent.  Did I miss something?

diegoreymendez replied:
What I was saying in my reply is that I think readability is more important than strict-consistency for this.

The reason why I added those spaces is that in my humble opinion long method headers followed by large blocks of text tent to be harder to read if they're all glued together.

This is easy to read because the method header is not long, and there isn't a long chunk of code after it:

```swift
func someFunc(a: ClassA, b: ClassB) {
    let a = someMethod()

    // etc...
}
```

Whereas this is harder to read because the method header is long, and there's a big block of text right after:

```swift
func someFunc(longerAName: ClassA, longerBName: ClassB, longerCName: ClassC,
    longerDName: ClassD, longerEName: ClassE) {
    let a = someLongMethod(withParameters: some, andParameters: someOthers)
    let b = someOtherLongMethod(havingParams: some, andHavingParams: someOthers)
    let c = makeAnotherCallWithoutManyParametersButWithALongName()

    // etc...
}
```

Regarding where the `{` is placed... the standard we have defines it goes on the same line as the method header - I don't see why an empty space would mean having to bring it to the next line.

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+        return [.Draft: draft,
+                .Scheduled: scheduled,
+                .Trashed: trashed,
+                .Published: published]
```
MUCH better than the original. Thank you =)

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+            return CGFloat(self.dynamicType.postCardRestoreCellRowHeight)
+        }
+        
+        var cell : PostCardTableViewCell!
```
No need for the `!` here.

diegoreymendez replied:
Since `cell` is initially `nil`, it needs to either be `?` or `!`.

In this case, the variable is `nil` when declared, but set to a value immediately after that, which is the [suggested used case](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/TheBasics.html#//apple_ref/doc/uid/TP40014097-CH5-ID309) for IUOs:

```Implicitly unwrapped optionals are useful when an optional’s value is confirmed to exist immediately after the optional is first defined and can definitely be assumed to exist at every point thereafter.```

A nicer option would be to create a method to return the correct cell, so that this doesn't need to be an IUO.  I didn't do this because I wanted to try to maintain parity for the sake of verifying the migration, but I can do it if you feel that's necessary.  The code above would become something like:

```swift
let cell = cellForPost(post)
```

aerych replied:
`Since cell is initially nil, it needs to either be ? or !.`

Actually, no.  :)  It just needs to be initialized before it is used.

diegoreymendez replied:
Sorry but I'm missing the idea.  Can you write the code we can use to avoid the `!`?

diegoreymendez replied:
You mean initialize it to one, and change it afterwards if needed?

aerych replied:
> You mean initialize it to one, and change it afterwards if needed?

Not really.  Try this in a play ground:
```
var foo: String
foo = "foo"
foo

var bar: String!
bar

var baz: String
baz
```
`baz` is going to throw an exception because it is used before it is initialized, but both `foo` and `bar` will be fine. `foo` works because we initialize the variable (assign it a value) before it is used. `bar` works because we make it an explicitly unwrapped optional which means it gets a value of `nil` from the start. 

When I said `No need for the ! here. ` its because we're assigning a value before its used and there is never a situation were it can be used before it has a value.  Hence no need for the optional.

diegoreymendez replied:
Interesting - didn't know if was possible to initialize after the declaration without the variable being an optional.

Thanks for the hint.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    private func createPostInNewEditor() {
+        let postViewController = WPPostViewController(draftForBlog: blog)
+        
+        postViewController.onClose = { [weak self] (viewController, changesSaved) -> () in
```
For cases where blocks return void ,the `-> ()` and ` -> Void`, its fine to just omit the return from the definition and a void type will be inferred.  Less noise.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+        let navController = UINavigationController(rootViewController: postViewController)
+        navController.restorationIdentifier = WPEditorNavigationRestorationID
+        navController.restorationClass = WPPostViewController.self
+        navController.toolbarHidden = false
```
It would be nice to keep the annotation from the original objc source as to why this is being set.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+        let identifier = NSStringFromClass(StatsPostDetailsTableViewController.self)
+        let service = BlogService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        let statsBundle = NSBundle(forClass: WPStatsViewController.self)
+        let path = statsBundle.pathForResource("WordPressCom-Stats-iOS", ofType: "bundle")!
```
It would be nice to avoid explicitly unwrapped optionals where ever possible. If for no other reason than paranoia. This way if a glitch is accidentally introduced in a future version of iOS or of the app we might avoid crashing on an unexpected nil.
Maybe change to something like 
```
        guard let path = statsBundle.pathForResource("WordPressCom-Stats-iOS", ofType: "bundle") else {
            return
        }
        let bundle = NSBundle(path: path)
        let statsStoryboard = UIStoryboard(name: self.dynamicType.statsStoryboardName, bundle: bundle)
        if let controller = statsStoryboard.instantiateViewControllerWithIdentifier(identifier) as? StatsPostDetailsTableViewController {
        
            controller.postID = apost.postID
            controller.postTitle = apost.titleForDisplay()
            controller.statsService = WPStatsService(siteId: blog.dotComID, siteTimeZone: service.timeZoneForBlog(blog), oauth2Token: blog.authToken, andCacheExpirationInterval: self.dynamicType.statsCacheInterval)
            
            navigationController?.pushViewController(controller, animated: true)
        }
```

diegoreymendez replied:
Some of these casts can be made safer easily - some others not so easily.  In this case I changed it to just do nothing in release if the stats VC cast fails, but assert in develop.  Please check out the new code.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    
+    // MARK: - Filter Related
+    
+    override internal func currentPostAuthorFilter() -> PostAuthorFilter {
```
No need to specify internal here.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+        return .Mine
+    }
+
+    override internal func setCurrentPostAuthorFilter(filter: PostAuthorFilter) {
```
No need to specify internal here.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    // MARK: - Cell Delegate Methods
+
+    func cell(cell: UITableViewCell!, receivedEditActionForProvider contentProvider: WPPostContentViewProvider!) {
+        let apost = contentProvider as! AbstractPost
```
I know that `contentProvier

aerych replied:
I know that `contentProvider` should only ever be an `AbstactPost` but for paranoia it would be nice to avoid the explicitly unwrapped optional. 
Can we change the method bodies in the delegate methods to something like: 
```
        if let apost = contentProvider as? AbstractPost {
            ...
        }
```

aerych replied:
Two suggestions about the PostCardTableViewCellDelegate methods.  

It would be nice to encapsulate the delete methods into an extension.  
```
extension PostListViewController: PostCardTableViewCellDelegate {
...
}
```

I also think its reasonable for this PR to edit `PostCardTableViewCellDelegate.h` and add `NS_ASSUME_NONNULL_BEGIN` `NS_ASSUME_NONNULL_END`.  This would let us avoid the `!` in the method definitions.  So
`func cell(cell: UITableViewCell!, receivedViewActionForProvider contentProvider: WPPostContentViewProvider!)`
becomes 
`func cell(cell: UITableViewCell, receivedViewActionForProvider contentProvider: WPPostContentViewProvider)`
Its just a bit tidier and as far as I can tell there is never a case where we're actually passing null to the delegate anyway.

diegoreymendez replied:
> I know that contentProvider should only ever be an AbstactPost but for paranoia it would be nice to avoid the explicitly unwrapped optional. 

I'd rather not have things that are clear mistakes on our end failing silently.  I think that's a recipe for not noticing logic errors.

The real issue in these lines of code is that we shouldn't need to cast `WPPostContentViewProvider` to `AbstractPost`.  It means something's not right - but fixing that is outside the scope of this PR (ie: convert 1:1).

> It would be nice to encapsulate the delete methods into an extension.

Will do.

> I also think its reasonable for this PR to edit PostCardTableViewCellDelegate.h and add `NS_ASSUME_NONNULL_BEGIN` `NS_ASSUME_NONNULL_END`.

Will do.

aerych replied:
> I'd rather not have things that are clear mistakes on our end failing silently. I think that's a recipe for not noticing logic errors.

Its a poor experience for the user, true. However a crash is a worse experience.  I'd use the optional and log it as an error so it can at least be surfaced in the activity log.

diegoreymendez replied:
> However a crash is a worse experience

The whole idea behind making the app fail is to make sure we don't code to allow invalid internal states that should not be there in the first place.  It's the same principle as with assertions.  You try to make it fail as soon and loudly as possible so it comes out during testing, as opposed to the log, which is much easier to ignore (intentionally or not), and more likely to make it to release.

diegoreymendez replied:
So, I spent some time thinking about this yesterday and today, and depending on the case I will take on of two actions:

1. In scenarios such as this one, in which the flow of execution would have no impact on the user experience, I will add an assertion for debug, a log for release, and interrupt the method execution.
2. In scenarios in which the invalid internal state would degrade user experience in other ways (such as seeing empty cells), I'll use `precondition()` instead of `assert()`, the rationale being that we really want to avoid invalid bug reports (ie: "I see empty cells") and instead get a crash and a precise stack trace.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+        guard let blogID = coder.decodeObjectForKey(postsViewControllerRestorationKey) as? String,
+            let objectURL = NSURL(string: blogID),
+            let objectID = context.persistentStoreCoordinator?.managedObjectIDForURIRepresentation(objectURL),
+            let restoredBlog = try? context.existingObjectWithID(objectID) as! Blog else {
```
Might as well change `as!` to `as?` here.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    // MARK: - UIViewController
+    
+    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
+        super.postListViewController = (segue.destinationViewController as! UITableViewController)
```
`super.postListViewController = segue.destinationViewController as? UITableViewController`  for paranoia.

diegoreymendez replied:
So... following up on what I mentioned [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/5156#discussion_r60729344) - this is a case in which allowing `super.postListViewController` to stay as `nil` will probably degrade the user experience in unexpected ways.

I added a `precondition()` call to make this explicit.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    // MARK: - Table View Handling
+    
+    private func tableView(tableView: UITableView, estimatedHeightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
+        let post = tableViewHandler?.resultsController.objectAtIndexPath(indexPath) as! Post
```
For paranoia:
```
        if let post = tableViewHandler?.resultsController.objectAtIndexPath(indexPath) as? Post {
            if cellIdentifierForPost(post) == self.dynamicType.postCardRestoreCellIdentifier {
                return CGFloat(self.dynamicType.postCardRestoreCellRowHeight)
            }
        }
        
        return CGFloat(self.dynamicType.postCardEstimatedRowHeight)
```

diegoreymendez replied:
This was marked as outdated due to the method visibility changing, but the comment is still alive...

Again, this is a case of the `resultsController` that should be retrieving `Post` objects and nothing else.

diegoreymendez replied:
Following up on [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/5156#discussion_r60729344), I'm adding an assertion, a log message and returning zero here.

In any case this will probably trigger a cell request which will cause a program interruption.

aerych commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,685 @@
+    }
+    
+    private func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let post = tableViewHandler?.resultsController.objectAtIndexPath(indexPath) as! Post
```
You already know what I'm going to say ;)

diegoreymendez replied:
Again, in this case... if we're ever retrieving something that's not a `Post`, I'd rather have this code fail so that we notice quickly.  There'd be no advantage in failing silently here - risking not noticing the problem.

aerych replied:
It would be better to prevent a crash for our user's sake.  We can always log the error and add asserts to help us catch any issue in debug/qa bulids.

diegoreymendez replied:
If this VC is trying to show anything other than posts, meaning there's some serious invalid state inside it - that's a good enough reason to shut down the app to prevent further data corruption.  There's no telling what else can break if we try to pretend nothing's wrong.

Also - there's no telling what kind of invalid bug reports we'll get if we try to pretend there's no issue here - as opposed to crashing right here, in which case the stack traces are going to show exactly what the problem is.

We should make sure we crash sooner rather than later, in a controlled way.  Invalid state is evil and should not be hidden.

diegoreymendez replied:
Actually this is a good case to talk about, since the method must return a `UITableViewCell` even if we don't have a post to show.

What would the cell show to the user?

My point is that for internal state errors (aka code bugs) I'd rather have the loud noise, but an exact and precise bug report with the stack trace.

I'd try to avoid at all costs a report saying something like: "I'm seeing some weird cells coming up in the list of posts showing an error message".

diegoreymendez replied:
I'm adding `preconditionFailure()` here if the post is not found.  Not having a requested post is an important case of invalid internal state.  We could return some empty cell here, but we'd basically be adding code that has no means to be tested.

aerych commented:
I'm getting a crash trying to view the post list: 
```
2016-04-20 10:37:01.874 WordPress[6306:49065] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'You must override configureCellsForLayout in a subclass'
*** First throw call stack:
```
It looks like maybe none of the configuration methods are visible to the `PostListViewController.swift` and they are not being overridden.

diegoreymendez commented:
I think I made the mistake of testing this PR before changing some methods to private.  My bad.  I'll re-review and get back to you @aerych.

diegoreymendez commented:
The crash on launch was fixed - it was apparently a result of some methods being changed to `private`.  Let me know if there's anything else.

diegoreymendez commented:
@aerych, @jleandroperez, @SergioEstevao - I changed quite a bit of code to improve error handling and fixed a small bug I found with stats.  This is ready for round 2.

jleandroperez commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,742 @@
+    static private let postsViewControllerRestorationKey = "PostsViewControllerRestorationKey"
+    static private let statsStoryboardName = "SiteStats"
+    static private let currentPostListStatusFilterKey = "CurrentPostListStatusFilterKey"
+    static private let currentPostAuthorFilterKey = "CurrentPostAuthorFilterKey"
```
Note for future PR's (no need to do anything here)

Maybe do some grouping like...

```
private enum CellNibs : String {
    case postCardTextCell = "PostCardTextCell"
    ....
}
```

diegoreymendez replied:
Yep - that's one of the improvements that can be done.  I avoided these kinds of things in this particular PR on purpose to make sure I'm not introducing new bugs and to make sure the code is easier to compare for the reviewers.

diegoreymendez commented:
@aerych, @jleandroperez, @SergioEstevao - It seems there's some discussion in this PR that belongs to a wider discussion that should IMHO be held outside of this particular PR.  It also seems the wider discussion is holding this PR back quite a bit, even though the new code is mostly equivalent to the original code.

The idea behind this PR is to offer a translation of the original code to Swift (nothing more, nothing less), so that once it's up we can make as many improvements as we see fit.  Focusing on making improvements here will only increase the size and complexity of this PR.

In my opinion we need to shift the focus from trying to make this code better, to trying to spot defects in the porting.

My 2 cents.

SergioEstevao commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,742 @@
+            return self.dynamicType.postCardRestoreCellRowHeight
+        }
+        
+        var cell : PostCardTableViewCell
```
Style nag: colon near the variable name like this: ```cell:```

SergioEstevao commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,742 @@
+    }
+    
+    private func cellIdentifierForPost(post: Post) -> String {
+        var identifier : String
```
Colon near the variable name.

SergioEstevao commented:
Clicking on the add post on the top right "+" icon on the post list is crashing the app with : ``` *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'You must override createPost in a subclass'
```

SergioEstevao commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,742 @@
+        
+        if recentlyTrashedPostObjectIDs?.containsObject(post.objectID) == true && currentPostListFilter()?.filterType != .Trashed {
+            identifier = self.dynamicType.postCardRestoreCellIdentifier
+        } else if post.pathForDisplayImage?.characters.count > 0 {
```
This logic is inverted from the original, if you have a path for a image you should display an image cell, if it's just text a text cell.

diegoreymendez replied:
Very good catch!

diegoreymendez commented:
@SergioEstevao - All comments addressed.

PS: the crash was actually because the `createPost` method existed but was `private`.  This is one of the reasons why I believe, calling methods like this from the superclass is a bad idea.  (no compile-time checks on if they're missing or not)

SergioEstevao commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -0,0 +1,742 @@
+        
+        var cell: PostCardTableViewCell
+        
+        if post.pathForDisplayImage?.characters.count > 0 {
```
@diegoreymendez  This logic is also inverted.

jleandroperez commented:
@diegoreymendez test cases (assigned to me) are all green!! great job sir!

jleandroperez commented:
@diegoreymendez sending a :shipit: on my side. Please, hold for @SergioEstevao and @aerych's squirrels 😄

diegoreymendez commented:
Ty @jleandroperez!

@SergioEstevao - Good catch again.  Ready for another look.

SergioEstevao commented:
@diegoreymendez :shipit: for me.

aerych commented:
@diegoreymendez :shipit: from me.

diegoreymendez commented:
Ty all!
